File: src/ruby/test/encoder_test.rb
# encoding: utf-8

require "test/unit"
require_relative "../lib/geodelta/encoder"

class GeoDeltaEncoderTest < Test::Unit::TestCase
  def setup
    @mod = GeoDelta::Encoder
  end

  def test_encode_world_delta
    assert_equal("Z", @mod.encode_world_delta(0))
    assert_equal("Y", @mod.encode_world_delta(1))
    assert_equal("X", @mod.encode_world_delta(2))
    assert_equal("W", @mod.encode_world_delta(3))
    assert_equal("V", @mod.encode_world_delta(4))
    assert_equal("T", @mod.encode_world_delta(5))
    assert_equal("S", @mod.encode_world_delta(6))
    assert_equal("R", @mod.encode_world_delta(7))
    assert_raise(RuntimeError) { @mod.encode_world_delta(-1) }
    assert_raise(RuntimeError) { @mod.encode_world_delta(8) }
  end

  def test_decode_world_delta
    assert_equal(0, @mod.decode_world_delta("Z"))
    assert_equal(1, @mod.decode_world_delta("Y"))
    assert_equal(2, @mod.decode_world_delta("X"))
    assert_equal(3, @mod.decode_world_delta("W"))
    assert_equal(4, @mod.decode_world_delta("V"))
    assert_equal(5, @mod.decode_world_delta("T"))
    assert_equal(6, @mod.decode_world_delta("S"))
    assert_equal(7, @mod.decode_world_delta("R"))
    assert_raise(RuntimeError) { @mod.decode_world_delta("z") }
    assert_raise(RuntimeError) { @mod.decode_world_delta("A") }
  end

  def test_encode_and_decode_world_delta
    (0..7).each { |id|
      encoded1 = @mod.encode_world_delta(id)
      decoded1 = @mod.decode_world_delta(encoded1)
      encoded2 = @mod.encode_world_delta(decoded1)
      assert_equal(id, decoded1)
      assert_equal(encoded1, encoded2)
    }
  end

  def test_encode_sub_delta__1
    assert_equal("2", @mod.encode_sub_delta([0, 0]))
    assert_equal("3", @mod.encode_sub_delta([0, 1]))
    assert_equal("4", @mod.encode_sub_delta([0, 2]))
    assert_equal("5", @mod.encode_sub_delta([0, 3]))
    assert_equal("6", @mod.encode_sub_delta([1, 0]))
    assert_equal("7", @mod.encode_sub_delta([1, 1]))
    assert_equal("8", @mod.encode_sub_delta([1, 2]))
    assert_equal("A", @mod.encode_sub_delta([1, 3]))
    assert_equal("B", @mod.encode_sub_delta([2, 0]))
    assert_equal("C", @mod.encode_sub_delta([2, 1]))
    assert_equal("D", @mod.encode_sub_delta([2, 2]))
    assert_equal("E", @mod.encode_sub_delta([2, 3]))
    assert_equal("F", @mod.encode_sub_delta([3, 0]))
    assert_equal("G", @mod.encode_sub_delta([3, 1]))
    assert_equal("H", @mod.encode_sub_delta([3, 2]))
    assert_equal("J", @mod.encode_sub_delta([3, 3]))
  end

  def test_encode_sub_delta__2
    assert_equal("K", @mod.encode_sub_delta([0]))
    assert_equal("M", @mod.encode_sub_delta([1]))
    assert_equal("N", @mod.encode_sub_delta([2]))
    assert_equal("P", @mod.encode_sub_delta([3]))
  end

  def test_encode_sub_delta__3
    assert_equal("2K", @mod.encode_sub_delta([0, 0, 0]))
    assert_equal("22", @mod.encode_sub_delta([0, 0, 0, 0]))
    assert_equal("3N", @mod.encode_sub_delta([0, 1, 2]))
    assert_equal("3E", @mod.encode_sub_delta([0, 1, 2, 3]))
  end

  def test_encode_sub_delta__4
    assert_raise(RuntimeError) { @mod.encode_sub_delta([]) }
    assert_raise(RuntimeError) { @mod.encode_sub_delta([-1]) }
    assert_raise(RuntimeError) { @mod.encode_sub_delta([4]) }
  end

  def test_decode_sub_delta__1
    assert_equal([0, 0], @mod.decode_sub_delta("2"))
    assert_equal([0, 1], @mod.decode_sub_delta("3"))
    assert_equal([0, 2], @mod.decode_sub_delta("4"))
    assert_equal([0, 3], @mod.decode_sub_delta("5"))
    assert_equal([1, 0], @mod.decode_sub_delta("6"))
    assert_equal([1, 1], @mod.decode_sub_delta("7"))
    assert_equal([1, 2], @mod.decode_sub_delta("8"))
    assert_equal([1, 3], @mod.decode_sub_delta("A"))
    assert_equal([2, 0], @mod.decode_sub_delta("B"))
    assert_equal([2, 1], @mod.decode_sub_delta("C"))
    assert_equal([2, 2], @mod.decode_sub_delta("D"))
    assert_equal([2, 3], @mod.decode_sub_delta("E"))
    assert_equal([3, 0], @mod.decode_sub_delta("F"))
    assert_equal([3, 1], @mod.decode_sub_delta("G"))
    assert_equal([3, 2], @mod.decode_sub_delta("H"))
    assert_equal([3, 3], @mod.decode_sub_delta("J"))
  end

  def test_decode_sub_delta__2
    assert_equal([0], @mod.decode_sub_delta("K"))
    assert_equal([1], @mod.decode_sub_delta("M"))
    assert_equal([2], @mod.decode_sub_delta("N"))
    assert_equal([3], @mod.decode_sub_delta("P"))
  end

  def test_decode_sub_delta__3
    assert_equal([0, 0, 0],    @mod.decode_sub_delta("2K"))
    assert_equal([0, 0, 0, 0], @mod.decode_sub_delta("22"))
    assert_equal([0, 1, 2],    @mod.decode_sub_delta("3N"))
    assert_equal([0, 1, 2, 3], @mod.decode_sub_delta("3E"))
  end

  def test_decode_sub_delta__4
    assert_raise(RuntimeError) { @mod.decode_sub_delta("") }
    assert_raise(RuntimeError) { @mod.decode_sub_delta("a") }
    assert_raise(RuntimeError) { @mod.decode_sub_delta("Z") }
  end

  def test_encode_and_decode_sub_delta__1
    (0..3).each { |id1|
      encoded1 = @mod.encode_sub_delta([id1])
      decoded1 = @mod.decode_sub_delta(encoded1)
      encoded2 = @mod.encode_sub_delta(decoded1)
      assert_equal([id1], decoded1)
      assert_equal(encoded1, encoded2)
    }
  end

  def test_encode_and_decode_sub_delta__2
    (0..3).each { |id1|
      (0..3).each { |id2|
        encoded1 = @mod.encode_sub_delta([id1, id2])
        decoded1 = @mod.decode_sub_delta(encoded1)
        encoded2 = @mod.encode_sub_delta(decoded1)
        assert_equal([id1, id2], decoded1)
        assert_equal(encoded1, encoded2)
      }
    }
  end

  def test_encode
    assert_equal("Z",   @mod.encode([0]))
    assert_equal("ZM",  @mod.encode([0, 1]))
    assert_equal("Z8",  @mod.encode([0, 1, 2]))
    assert_equal("Z8P", @mod.encode([0, 1, 2, 3]))
    assert_equal("R",   @mod.encode([7]))
    assert_equal("RP",  @mod.encode([7, 3]))
    assert_equal("RH",  @mod.encode([7, 3, 2]))
    assert_equal("RHM", @mod.encode([7, 3, 2, 1]))
    assert_raise(RuntimeError) { @mod.encode([]) }
  end

  def test_decode
    assert_equal([0],          @mod.decode("Z"))
    assert_equal([0, 1],       @mod.decode("ZM"))
    assert_equal([0, 1, 2],    @mod.decode("Z8"))
    assert_equal([0, 1, 2, 3], @mod.decode("Z8P"))
    assert_equal([7],          @mod.decode("R"))
    assert_equal([7, 3],       @mod.decode("RP"))
    assert_equal([7, 3, 2],    @mod.decode("RH"))
    assert_equal([7, 3, 2, 1], @mod.decode("RHM"))
    assert_raise(RuntimeError) { @mod.encode("") }
  end

  def test_encode_and_decode__rush
    world = (0..7).to_a
    sub   = (0..3).to_a
    1000.times {
      ids = [world[rand(world.size)]] + rand(20).times.map { sub[rand(sub.size)] }
      encoded1 = @mod.encode(ids)
      decoded1 = @mod.decode(encoded1)
      encoded2 = @mod.encode(decoded1)
      assert_equal(ids, decoded1)
      assert_equal(encoded1, encoded2)
    }
  end
end

File: src/js/test/encoder_test.js

function test_namespace() {
  assertNotUndefined(geodelta);
  assertNotUndefined(geodelta.encoder);
}

function test_encodeWorldDelta()
{
  assertEquals("Z", geodelta.encoder.encodeWorldDelta(0));
  assertEquals("Y", geodelta.encoder.encodeWorldDelta(1));
  assertEquals("X", geodelta.encoder.encodeWorldDelta(2));
  assertEquals("W", geodelta.encoder.encodeWorldDelta(3));
  assertEquals("V", geodelta.encoder.encodeWorldDelta(4));
  assertEquals("T", geodelta.encoder.encodeWorldDelta(5));
  assertEquals("S", geodelta.encoder.encodeWorldDelta(6));
  assertEquals("R", geodelta.encoder.encodeWorldDelta(7));
}

function test_encodeWorldDelta__invalidArg1()
{
  try
  {
    geodelta.encoder.encodeWorldDelta(-1);
  }
  catch ( e )
  {
    return;
  }
  fail("exception not raised");
}

function test_encodeWorldDelta__invalidArg2()
{
  try
  {
    geodelta.encoder.encodeWorldDelta(8);
  }
  catch ( e )
  {
    return;
  }
  fail("exception not raised");
}

function test_decodeWorldDelta()
{
  assertEquals(0, geodelta.encoder.decodeWorldDelta("Z"));
  assertEquals(1, geodelta.encoder.decodeWorldDelta("Y"));
  assertEquals(2, geodelta.encoder.decodeWorldDelta("X"));
  assertEquals(3, geodelta.encoder.decodeWorldDelta("W"));
  assertEquals(4, geodelta.encoder.decodeWorldDelta("V"));
  assertEquals(5, geodelta.encoder.decodeWorldDelta("T"));
  assertEquals(6, geodelta.encoder.decodeWorldDelta("S"));
  assertEquals(7, geodelta.encoder.decodeWorldDelta("R"));
}

/*
    @Test(expected = IllegalArgumentException.class)
    public void decodeWorldDelta__invalidArg1()
    {
        geodelta.encoder.decodeWorldDelta("z");
    }

    @Test(expected = IllegalArgumentException.class)
    public void decodeWorldDelta__invalidArg2()
    {
        geodelta.encoder.decodeWorldDelta("A");
    }

    @Test
    public void allEncodeAndDecodeWorldDelta()
    {
        for ( int id = 0; id <= 7; id++ )
        {
            final char encoded1 = geodelta.encoder.encodeWorldDelta((byte)id);
            final byte decoded1 = geodelta.encoder.decodeWorldDelta(encoded1);
            final char encoded2 = geodelta.encoder.encodeWorldDelta(decoded1);
            assertEquals(encoded1, encoded2);
        }
    }
*/

function test_encodeSubDelta__1()
{
  assertEquals("2", geodelta.encoder.encodeSubDelta([0, 0]));
  assertEquals("3", geodelta.encoder.encodeSubDelta([0, 1]));
  assertEquals("4", geodelta.encoder.encodeSubDelta([0, 2]));
  assertEquals("5", geodelta.encoder.encodeSubDelta([0, 3]));
  assertEquals("6", geodelta.encoder.encodeSubDelta([1, 0]));
  assertEquals("7", geodelta.encoder.encodeSubDelta([1, 1]));
  assertEquals("8", geodelta.encoder.encodeSubDelta([1, 2]));
  assertEquals("A", geodelta.encoder.encodeSubDelta([1, 3]));
  assertEquals("B", geodelta.encoder.encodeSubDelta([2, 0]));
  assertEquals("C", geodelta.encoder.encodeSubDelta([2, 1]));
  assertEquals("D", geodelta.encoder.encodeSubDelta([2, 2]));
  assertEquals("E", geodelta.encoder.encodeSubDelta([2, 3]));
  assertEquals("F", geodelta.encoder.encodeSubDelta([3, 0]));
  assertEquals("G", geodelta.encoder.encodeSubDelta([3, 1]));
  assertEquals("H", geodelta.encoder.encodeSubDelta([3, 2]));
  assertEquals("J", geodelta.encoder.encodeSubDelta([3, 3]));
}

function test_encodeSubDelta__2()
{
  assertEquals("K", geodelta.encoder.encodeSubDelta([0]));
  assertEquals("M", geodelta.encoder.encodeSubDelta([1]));
  assertEquals("N", geodelta.encoder.encodeSubDelta([2]));
  assertEquals("P", geodelta.encoder.encodeSubDelta([3]));
}

function test_encodeSubDelta__3()
{
  assertEquals("2K", geodelta.encoder.encodeSubDelta([0, 0, 0]));
  assertEquals("22", geodelta.encoder.encodeSubDelta([0, 0, 0, 0]));
  assertEquals("3N", geodelta.encoder.encodeSubDelta([0, 1, 2]));
  assertEquals("3E", geodelta.encoder.encodeSubDelta([0, 1, 2, 3]));
}

/*
    @Test(expected = IllegalArgumentException.class)
    public void encodeSubDelta__invalidArg1()
    {
        geodelta.encoder.encodeSubDelta(null);
    }

    @Test(expected = IllegalArgumentException.class)
    public void encodeSubDelta__invalidArg2()
    {
        geodelta.encoder.encodeSubDelta(new byte[0]);
    }

    @Test(expected = IllegalArgumentException.class)
    public void encodeSubDelta__invalidArg3()
    {
        geodelta.encoder.encodeSubDelta(new byte[] {-1});
    }

    @Test(expected = IllegalArgumentException.class)
    public void encodeSubDelta__invalidArg4()
    {
        geodelta.encoder.encodeSubDelta(new byte[] {4});
    }
*/

function test_decodeSubDelta__1()
{
  assertArrayEquals([0, 0], geodelta.encoder.decodeSubDelta("2"));
  assertArrayEquals([0, 1], geodelta.encoder.decodeSubDelta("3"));
  assertArrayEquals([0, 2], geodelta.encoder.decodeSubDelta("4"));
  assertArrayEquals([0, 3], geodelta.encoder.decodeSubDelta("5"));
  assertArrayEquals([1, 0], geodelta.encoder.decodeSubDelta("6"));
  assertArrayEquals([1, 1], geodelta.encoder.decodeSubDelta("7"));
  assertArrayEquals([1, 2], geodelta.encoder.decodeSubDelta("8"));
  assertArrayEquals([1, 3], geodelta.encoder.decodeSubDelta("A"));
  assertArrayEquals([2, 0], geodelta.encoder.decodeSubDelta("B"));
  assertArrayEquals([2, 1], geodelta.encoder.decodeSubDelta("C"));
  assertArrayEquals([2, 2], geodelta.encoder.decodeSubDelta("D"));
  assertArrayEquals([2, 3], geodelta.encoder.decodeSubDelta("E"));
  assertArrayEquals([3, 0], geodelta.encoder.decodeSubDelta("F"));
  assertArrayEquals([3, 1], geodelta.encoder.decodeSubDelta("G"));
  assertArrayEquals([3, 2], geodelta.encoder.decodeSubDelta("H"));
  assertArrayEquals([3, 3], geodelta.encoder.decodeSubDelta("J"));
}

function test_decodeSubDelta__2()
{
  assertArrayEquals([0], geodelta.encoder.decodeSubDelta("K"));
  assertArrayEquals([1], geodelta.encoder.decodeSubDelta("M"));
  assertArrayEquals([2], geodelta.encoder.decodeSubDelta("N"));
  assertArrayEquals([3], geodelta.encoder.decodeSubDelta("P"));
}

function test_decodeSubDelta__3()
{
  assertArrayEquals([0, 0, 0], geodelta.encoder.decodeSubDelta("2K"));
  assertArrayEquals([0, 0, 0, 0], geodelta.encoder.decodeSubDelta("22"));
  assertArrayEquals([0, 1, 2], geodelta.encoder.decodeSubDelta("3N"));
  assertArrayEquals([0, 1, 2, 3], geodelta.encoder.decodeSubDelta("3E"));
}

/*
    @Test(expected = IllegalArgumentException.class)
    public void decodeSubDelta__invalidArg1()
    {
        geodelta.encoder.decodeSubDelta(null);
    }

    @Test(expected = IllegalArgumentException.class)
    public void decodeSubDelta__invalidArg2()
    {
        geodelta.encoder.decodeSubDelta("");
    }

    @Test(expected = IllegalArgumentException.class)
    public void decodeSubDelta__invalidArg3()
    {
        geodelta.encoder.decodeSubDelta("1");
    }

    @Test(expected = IllegalArgumentException.class)
    public void decodeSubDelta__invalidArg4()
    {
        geodelta.encoder.decodeSubDelta("Z");
    }

    @Test
    public void allEncodeAndDecodeSubDelta__level2()
    {
        for ( int id1 = 0; id1 <= 3; id1++ )
        {
            final byte[] ids = {(byte)id1};
            final String encoded1 = geodelta.encoder.encodeSubDelta(ids);
            final byte[] decoded1 = geodelta.encoder.decodeSubDelta(encoded1);
            final String encoded2 = geodelta.encoder.encodeSubDelta(decoded1);
            assertEquals(encoded1, encoded2);
        }
    }

    @Test
    public void allEncodeAndDecodeSubDelta__level3()
    {
        for ( int id1 = 0; id1 <= 3; id1++ )
        {
            for ( int id2 = 0; id2 <= 3; id2++ )
            {
                final byte[] ids = {(byte)id1, (byte)id2};
                final String encoded1 = geodelta.encoder.encodeSubDelta(ids);
                final byte[] decoded1 = geodelta.encoder.decodeSubDelta(encoded1);
                final String encoded2 = geodelta.encoder.encodeSubDelta(decoded1);
                assertEquals(encoded1, encoded2);
            }
        }
    }
*/

function test_encode()
{
  assertEquals("Z",   geodelta.encoder.encode([0]));
  assertEquals("ZM",  geodelta.encoder.encode([0, 1]));
  assertEquals("Z8",  geodelta.encoder.encode([0, 1, 2]));
  assertEquals("Z8P", geodelta.encoder.encode([0, 1, 2, 3]));
  assertEquals("R",   geodelta.encoder.encode([7]));
  assertEquals("RP",  geodelta.encoder.encode([7, 3]));
  assertEquals("RH",  geodelta.encoder.encode([7, 3, 2]));
  assertEquals("RHM", geodelta.encoder.encode([7, 3, 2, 1]));
}

/*
    @Test(expected = IllegalArgumentException.class)
    public void encode__invalidArg1()
    {
        geodelta.encoder.encode(null);
    }

    @Test(expected = IllegalArgumentException.class)
    public void encode__invalidArg2()
    {
        geodelta.encoder.encode(new byte[0]);
    }
*/

function test_decode()
{
  assertArrayEquals([0],          geodelta.encoder.decode("Z"));
  assertArrayEquals([0, 1],       geodelta.encoder.decode("ZM"));
  assertArrayEquals([0, 1, 2],    geodelta.encoder.decode("Z8"));
  assertArrayEquals([0, 1, 2, 3], geodelta.encoder.decode("Z8P"));
  assertArrayEquals([7],          geodelta.encoder.decode("R"));
  assertArrayEquals([7, 3],       geodelta.encoder.decode("RP"));
  assertArrayEquals([7, 3, 2],    geodelta.encoder.decode("RH"));
  assertArrayEquals([7, 3, 2, 1], geodelta.encoder.decode("RHM"));
}

/*
    @Test(expected = IllegalArgumentException.class)
    public void decode__invalidArg1()
    {
        geodelta.encoder.decode(null);
    }

    @Test(expected = IllegalArgumentException.class)
    public void decode__invalidArg2()
    {
        geodelta.encoder.decode("");
    }

    @Test
    public void randomEncodeAndDecode()
    {
        final Random r = new Random();

        for ( int i = 0; i < 1000; i++ )
        {
            final byte[] ids = createRandomDeltaIds(r, 20);
            final String encoded1 = geodelta.encoder.encode(ids);
            final byte[] decoded1 = geodelta.encoder.decode(encoded1);
            final String encoded2 = geodelta.encoder.encode(decoded1);
            assertArrayEquals(ids, decoded1);
            assertEquals(encoded1, encoded2);
        }
    }

    // FIXME: refactoring
    private byte[] createRandomDeltaIds(final Random random, final int level)
    {
        final byte[] ids = new byte[random.nextInt(level) + 1];
        ids[0] = (byte)random.nextInt(8);
        for ( int i = 1; i < ids.length; i++ )
        {
            ids[i] = (byte)random.nextInt(4);
        }
        return ids;
    }
*/
